packages/react-components/src/components/table-output-component.tsx

```diff
@@ -263,15 +414,48 @@ export class TableOutputComponent extends AbstractOutputComponent<TableOutputPro
     }
 
     private onModelUpdated = async () => {
-        if (this.pendingIndex && this.gridApi) {
-            if (this.gridApi.getSelectedNodes().length === 0) {
-                this.gridApi.getDisplayedRowAtIndex(this.pendingIndex).setSelected(true);
-            }
-            this.pendingIndex = undefined;
-        }
+        this.selectRows();
+
         if (this.columnArray.length > 0 && !this.columnsPacked && this.columnApi) {
             this.columnApi.autoSizeAllColumns();
             this.columnsPacked = true;
         }
     };
+
+    private isValidRowSelection(rowNode: RowNode): boolean {
+        if ((this.enableIndexSelection && rowNode.rowIndex >= Math.min(this.selectStartIndex, this.selectEndIndex)
+            && rowNode.rowIndex <= Math.max(this.selectStartIndex, this.selectEndIndex)) || (!this.enableIndexSelection
+                && this.timestampCol && rowNode.data[this.timestampCol] >= Math.min(this.startTimestamp, this.endTimestamp)
+                && rowNode.data[this.timestampCol] <= Math.max(this.startTimestamp, this.endTimestamp))) {
+            return true;
+        }
+        return false;
+    }
+
+    private selectRows(): void {
+        let isNodeSelected = false;
+        if (this.gridApi) {
+            this.gridApi.forEachNode(rowNode => {
+                if (rowNode.id) {
+                    if (this.isValidRowSelection(rowNode)) {
+                        isNodeSelected = true;
+                        rowNode.setSelected(true);
+                    } else {
+                        rowNode.setSelected(false);
+                    }
+                    if (!this.enableIndexSelection) {
+                        if (this.endTimestamp < this.startTimestamp) {
+                            this.selectEndIndex = Math.min(this.selectEndIndex, rowNode.rowIndex);
+                        } else {
+                            this.selectEndIndex = Math.max(this.selectEndIndex, rowNode.rowIndex);
+                        }
+                    }
+                }
+            });
+
+            if (this.gridApi.getSelectedNodes.length === 0 && !this.enableIndexSelection && isNodeSelected === false) {
+                this.selectStartIndex = this.selectEndIndex = -1;
```

Comment:
The same problem still happens. See the log below.

I think this line is the problem. I don't think it's the job of selectRows() to clear the start/end index when enableIndexSelection is false.

I believe the intention is to have correct start/end index in case the user extends a time-based selection into an index-based selection, by shift-clicking on the event table after having selected a time range in a time graph.

In that case I believe that the right place to clear the start/end index is at the reception of a new time range in handleTimeSelectionRange(), but only when the start time has changed e.g. in the branch where it fetches the new start index, just before calling fetchTableIndex().

Another issue, above it is updating the end index when using time range selection. I believe that in some cases it also needs to set the start index (e.g. if it has never been set following a time range change). As far as I can tell, the effect of not setting it is that, after a time range selection, you can extend selection end by shift+clicking the table on an unselected event below the current multiple selection, but it does not work if you try to extend the selection start by shift-clicking the table on an unselected event above the current multiple selection.

root INFO *** handleTimeSelectionChange({"start":1341337214,"end":1341337214})
root INFO vvv fetching table index @4738416805
root INFO *** handleTimeSelectionChange({"start":1341337214,"end":1341545475})
root INFO selectRows() enableIndexSelection=false selectStartIndex=80 selectEndIndex=80
root INFO *** setting selectStartIndex = selectEndIndex = -1
root INFO *** handleTimeSelectionChange({"start":1341337214,"end":1341566293})
root INFO selectRows() enableIndexSelection=false selectStartIndex=-1 selectEndIndex=-1
root INFO *** setting selectStartIndex = selectEndIndex = -1
root INFO *** handleTimeSelectionChange({"start":1341337214,"end":1342005891})
root INFO selectRows() enableIndexSelection=false selectStartIndex=-1 selectEndIndex=-1
root INFO *** setting selectStartIndex = selectEndIndex = -1
root INFO ^^^ fetching table index returns 4269
root INFO handleTimeSelectionChange() setting selectStartIndex = selectEndIndex = 4269
root INFO selectRows() enableIndexSelection=false selectStartIndex=4269 selectEndIndex=4269
root INFO *** setting selectStartIndex = selectEndIndex = -1
root INFO vvv fetching table lines @4200
root INFO *** handleTimeSelectionChange({"start":1341337214,"end":1341566293})
root INFO selectRows() enableIndexSelection=false selectStartIndex=-1 selectEndIndex=-1
root INFO *** setting selectStartIndex = selectEndIndex = -1
root INFO *** handleTimeSelectionChange({"start":1341337214,"end":1341545475})
root INFO selectRows() enableIndexSelection=false selectStartIndex=-1 selectEndIndex=-1
root INFO *** setting selectStartIndex = selectEndIndex = -1
root INFO *** handleTimeSelectionChange({"start":1341337214,"end":1341337214})
root INFO selectRows() enableIndexSelection=true selectStartIndex=-1 selectEndIndex=-1
root INFO ^^^ fetching table lines returns
root INFO selectRows() enableIndexSelection=true selectStartIndex=-1 selectEndIndex=-1

Reply:
If you look at the log above, when fetchTableIndex() returns the selection start/end is set to 4269. But then when the time range selection is selected, the start/end is cleared to -1 because the table currently doesn't have a selection. So when the time range selection comes back to a single time and index-based selection is set, the start/end has been lost.

I also wonder, in selectRows() there is code to increase the end index when the range selection end time is increased, but it does not decrease the end index when the time range selection end time is decreased. For example, extending from [10, 20] to [10, 30] then back to [10, 25], the end index is stuck at 30. So if we really need the start/end index to be updated there, it will not work properly.

Also the updating of end index in selectRows() should only happen for selected row nodes! Otherwise it updates the end index just because a row exists, regardless of whether it is part of the selection. See example below, I was extending time selection to cover [1, 1] then [1, 2] then [1, 1] again, but end index becomes 199.

root INFO vvv fetching table lines @0
root INFO *** handleTimeSelectionChange({"start":11.833151488666372,"end":11.833151488666372})
root INFO vvv fetching table index @3397079602
root INFO ^^^ fetching table lines returns
root INFO selectRows() enableIndexSelection=true selectStartIndex=-1 selectEndIndex=-1
root INFO ^^^ fetching table index returns 1
root INFO handleTimeSelectionChange() setting selectStartIndex = selectEndIndex = 1
root INFO selectRows() enableIndexSelection=true selectStartIndex=1 selectEndIndex=1
root INFO setSelected(true) #1
root INFO *** handleTimeSelectionChange({"start":11.833151488666372,"end":747.4607357007592})
root INFO selectRows() enableIndexSelection=false selectStartIndex=1 selectEndIndex=1
root INFO setSelected(true) #1
root INFO setSelected(true) #2
root INFO *** handleTimeSelectionChange({"start":11.833151488666372,"end":464.4511959301551})
root INFO selectRows() enableIndexSelection=false selectStartIndex=1 selectEndIndex=199
root INFO setSelected(true) #1